File: homework1/Python/hsinyu_chang_task1.py
import sys
import json
from pyspark import SparkContext, SparkConf, StorageLevel
from operator import add

#get arguments 
review = sys.argv[1]
outfile = sys.argv[2]

#construct configuration
conf = SparkConf()
conf.setMaster("local")
conf.setAppName("hw1")
sc = SparkContext.getOrCreate(conf)

#read jsonfile and parse it
reviews = sc.textFile(review)
reviewsRDD = reviews.map(lambda line: json.loads(line))\
.map(lambda review: (review['date'], review['user_id'], review['business_id'])).persist(StorageLevel.DISK_ONLY)  

#a
a = reviewsRDD.count()
#b
b = reviewsRDD.filter(lambda review: review[0][:4] == '2018').count()

#c, d
temp = reviewsRDD.map(lambda review: (review[1], 1)).reduceByKey(add).persist(StorageLevel.DISK_ONLY) 
c = temp.count()
d = temp.takeOrdered(10, key=lambda user: (-user[1], user[0]))
d = [list(r) for r in d]

temp.unpersist()

#e, f
temp = reviewsRDD.map(lambda review: (review[2], 1)).reduceByKey(add).persist(StorageLevel.DISK_ONLY) 
e = temp.count()
f = temp.takeOrdered(10, key=lambda business: (-business[1], business[0]))
f = [list(r) for r in f]

temp.unpersist()

#output dictionary
task1 = {
    "n_review" : a,
    "n_review_2018" : b,
    "n_user" : c,
    "top10_user": d,
    "n_business" : e,
    "top10_business": f
}

#write output json file
with open(outfile, 'w') as fd:
    json.dump(task1, fd)

File: homework1/Scala/hsinyu_chang_task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.storage.StorageLevel
import org.json4s._
//import org.json4s.JsonDSL._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization.writePretty
import java.io._

object hsinyu_chang_task1 {

    implicit val formats = DefaultFormats

    def main(args: Array[String]) {

        val inputFile = args(0)
        val outputFile = args(1)

        val conf = new SparkConf()

        conf.setMaster("local[*]")
        conf.setAppName("INF553")
        conf.set("spark.executor.memory", "4g")
        conf.set("spark.driver.memory", "4g")

        val sc = new SparkContext(conf)
        val reviews = sc.textFile(inputFile)

        val reviewsRDD = reviews.map(line => parse(line).extract[Map[String, String]]).
          map(review => (review("date"), review("user_id"), review("business_id"))).
          persist(StorageLevel.DISK_ONLY)

        val a = reviewsRDD.count()
        val b = reviewsRDD.filter(review => review._1.slice(0, 4) == "2018").count()


        val d = reviewsRDD.map(review => (review._2, 1)).reduceByKey(_ + _).
          sortBy(user => (-user._2, user._1)).
          persist(StorageLevel.DISK_ONLY)

        val c = d.count()
        val outputd =  d.map(user => List(user._1,user._2)).take(10)


        val f = reviewsRDD.map(review => (review._3, 1)).reduceByKey(_ + _).
          sortBy(business => (-business._2, business._1)).
          persist(StorageLevel.DISK_ONLY)

        val e = f.count()
        val outputf =  f.map(business => List(business._1,business._2)).take(10)


        val json = Map(
            "n_review" -> a,
            "n_review_2018" -> b,
            "n_user" -> c,
            "top10_user" -> outputd ,
            "n_business" -> e,
            "top10_business" -> outputf)


        val writer = new PrintWriter(new File(outputFile))
        writer.write(writePretty(json))
        writer.close()
    }
}
